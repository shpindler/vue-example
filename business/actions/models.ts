import { RawAction } from './types'

export class Action {
  url?: string;
  name: string;
  icon?: string;
  actions?: Action[];
  params?: { [key: string]: any };

  constructor(data: Action) {
    this.url = data.url
    this.name = data.name
    this.icon = data.icon
    this.actions = data.actions
    this.params = data.params
  }

  static parse(data: RawAction): Action {
    return new Action({
      name: data.text,
      icon: data.icon,
      url: data.url,
      actions: data.items ? data.items.map(Action.parse) : [],
      params: data.context,
    })
  }
}
