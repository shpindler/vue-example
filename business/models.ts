export class Color {
  red!: number
  green!: number
  blue!: number

  constructor(red: number, green: number, blue: number) {
    this.red = red
    this.green = green
    this.blue = blue
  }

  toString(): string {
    return `rgb(${this.red}, ${this.green}, ${this.blue})`
  }

  rgba(a: number): string {
    return `rgba(${this.red}, ${this.green}, ${this.blue}, ${a})`
  }
}
