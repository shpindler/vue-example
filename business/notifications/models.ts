import { Color } from '../models'
import { NotificationTypes, RawNotification } from './types'

export abstract class Notification {
  abstract readonly type: NotificationTypes
  abstract readonly image: string
  abstract readonly color: Color
  readonly title: string
  readonly text: string
  readonly date: Date
  seen: boolean

  constructor(data: Pick<Notification, 'title' | 'text' | 'date' | 'seen'>) {
    this.title = data.title
    this.text = data.text
    this.date = data.date
    this.seen = data.seen
  }
}

export class NegativeNotification extends Notification {
  type = NotificationTypes.NEGATIVE
  image = 'error'
  color = new Color(255, 192, 33)
}

export class PositiveNotification extends Notification {
  type = NotificationTypes.POSITIVE
  image = 'check_circle'
  color = new Color(71, 215, 100)
}

export class NeutralNotification extends Notification {
  type = NotificationTypes.NEUTRAL
  image = 'info'
  color = new Color(47, 134, 235)
}

export function parseNotification(data: RawNotification): Notification {
  const preparedData = {
    title: data.header,
    text: data.text,
    date: new Date(data.created),
    seen: data.seen,
  }
  switch (data.type) {
    case NotificationTypes.NEGATIVE:
      return new NegativeNotification(preparedData)
    case NotificationTypes.POSITIVE:
      return new PositiveNotification(preparedData)
    case NotificationTypes.NEUTRAL:
      return new NeutralNotification(preparedData)
  }
}
