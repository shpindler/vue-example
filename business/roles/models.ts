import { RawRole } from './types'

export class Role {
  id!: number;
  name!: string;
  userType!: string;

  constructor(role: Role) {
    this.id = role.id
    this.name = role.name
    this.userType = role.userType
  }

  static parse(role: RawRole): Role {
    return new Role({
      id: role.userRoleId,
      name: role.userRoleName,
      userType: role.userType,
    })
  }
}
