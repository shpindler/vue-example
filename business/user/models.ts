export class User {
  firstName = ''
  lastName = ''
  avatar = ''
  position = ''

  constructor({
    firstName = '',
    lastName = '',
    avatar = '',
    position = '',
  }: Partial<User>) {
    this.firstName = firstName
    this.lastName = lastName
    this.avatar = avatar
    this.position = position
  }

  get fullName(): string {
    return `${this.firstName} ${this.lastName}`
  }
}

export class UserProfile extends User {
  patronymic = ''
  email = ''
  ucpk = ''
  phone = ''
  _dateOfBirth = ''
  position = ''
  isNewUser = false

  constructor(user: Partial<UserProfile>) {
    super(user)
    this.patronymic = user.patronymic || ''
    this.email = user.email || ''
    this.ucpk = user.ucpk || ''
    this.phone = user.phone || ''
    this._dateOfBirth = user.dateOfBirth as unknown as string || ''
    this.position = user.position || ''
    this.isNewUser = user.isNewUser || false
  }

  get dateOfBirth(): Date | undefined {
    if (!this._dateOfBirth) {
      return undefined
    }
    const [year, month, day] = this._dateOfBirth.split('-').map(p => parseInt(p))
    return new Date(year, month - 1, day)
  }
}
