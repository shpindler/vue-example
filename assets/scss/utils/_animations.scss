@keyframes slide-to-left {
  0% {
    transform: translateX(0);
  }

  100% {
    transform: translateX(-100%);
  }
}

@keyframes slide-to-right {
  0% {
    transform: translateX(-100%);
  }

  100% {
    transform: translateX(0);
  }
}

@keyframes fade-in {
  0% {
    opacity: 0;
  }

  100% {
    opacity: 1;
  }
}

@keyframes increase {
   from { left: -5%; width: 5%; }
   to { left: 130%; width: 100%; }
}

@keyframes decrease {
   from { left: -80%; width: 80%; }
   to { left: 110%; width: 10%; }
}

.slide {
  &-to {
    @each $direction in (left, right) {
      &-#{$direction} {
        animation-name: slide-to-#{$direction};
        animation-fill-mode: forwards;
        animation-duration: 0.6s;
      }
    }
  }
}

.fade-in {
  animation-name: fade-in;
  animation-fill-mode: forwards;
  animation-duration: 0.4s;
}

.fade-enter-active,
.fade-leave-active {
  transition: opacity 0.3s;
}

.fade-enter,
.fade-leave-to {
  opacity: 0;
}

.slide-x-enter-active {
  animation: slide-to-right 0.6s ease;
}

.slide-x-leave-active {
  animation: slide-to-left 0.6s ease;
}

.inc {
  animation: increase 2s infinite;
}

.dec {
  animation: decrease 2s 0.5s infinite;
}

.expand-enter-active,
.expand-leave-active {
  transition: height 0.2s ease-in-out;
  overflow: hidden;
}

.expand-enter,
.expand-leave-to {
  height: 0;
}
