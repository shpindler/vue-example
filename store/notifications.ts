export { NotificationsModule as default } from '@/business/notifications/store'
