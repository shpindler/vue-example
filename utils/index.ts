interface Transformable {
  [key: string]: any;
}

/** Преобразует ключи. */
export function transformKeys(
  transformer: (key: string) => string,
  data: Transformable,
): Transformable {
  for (const key in data) {
    if (Object.prototype.hasOwnProperty.call(data, key)) {
      let value = data[key]
      if (typeof data[key] === 'object') {
        value = transformKeys(transformer, value)
      } else if (Array.isArray(data[key])) {
        value = value.map((item: Transformable) => {
          return transformKeys(transformer, item)
        })
      }
      const newKey = transformer(key)
      if (newKey !== key) {
        data[transformer(key)] = value
        delete data[key]
      }
    }
  }

  return data
}

export function getCookie(name: string): string {
  let cookieValue = ''
  if (document.cookie && document.cookie !== '') {
    const cookies = document.cookie.split(';')
    for (let i = 0; i < cookies.length; i++) {
      const cookie = cookies[i].trim()
      if (cookie.substring(0, name.length + 1) === name + '=') {
        cookieValue = decodeURIComponent(cookie.substring(name.length + 1))
        break
      }
    }
  }
  return cookieValue
}

export function convertToFormData(obj: Transformable): FormData {
  const formData = new FormData()
  for (const key in obj) {
    if (Object.prototype.hasOwnProperty.call(obj, key)) {
      formData.append(key, obj[key])
    }
  }
  return formData
}

export function sliceStartSlash(str: string): string {
  if (str[0] === '/') {
    return str.slice(1, str.length)
  }
  return str
}

export function sliceEndSlash(str: string): string {
  if (str[str.length - 1] === '/') {
    return str.slice(0, str.length - 1)
  }
  return str
}

export function decline(
  variants: [string, string, string],
  count: number,
): string {
  if (count % 10 === 1) {
    return variants[0]
  } else if (count <= 4 && count >= 2) {
    return variants[1]
  }
  return variants[2]
}

export function formatDates(data: Transformable): Transformable {
  for (const key in data) {
    if (Object.prototype.hasOwnProperty.call(data, key)) {
      if (data[key] instanceof Date) {
        const [day, month, year] = data[key].toLocaleDateString('ru').split('.')
        data[key] = `${year}-${month}-${day}`
      }
    }
  }
  return data
}

export function removeUndefined(data: Transformable): Transformable {
  for (const key in data) {
    if (Object.prototype.hasOwnProperty.call(data, key)) {
      if (data[key] === undefined) {
        delete data[key]
      }
    }
  }
  return data
}

export function formatDate(value: Date): string {
  return value.toLocaleDateString('ru')
}
