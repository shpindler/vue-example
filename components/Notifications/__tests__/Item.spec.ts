import { createLocalVue, shallowMount } from '@vue/test-utils'
import Component from '../Item.vue'
import { NegativeNotification } from '@/business/notifications/models'
import { formatDate } from '@/utils'

const localVue = createLocalVue()
localVue.filter('date', formatDate)

const notification = new NegativeNotification({
  title: '22',
  text: 'Тестовое уведомление.',
  date: new Date('2000-10-10'),
  seen: true,
})

const unseenNotification = new NegativeNotification({
  title: '22',
  text: 'Тестовое уведомление.',
  date: new Date('2000-10-10'),
  seen: false,
})

const stubs = {
  'vue-row': true,
  'vue-col': true,
  'vue-icon': true,
}

describe('Notification', () => {
  test('является экземпляром Vue.', () => {
    const wrapper = shallowMount(Component, {
      localVue,
      propsData: {
        notification,
      },
      stubs,
    })
    expect(wrapper.isVueInstance()).toBeTruthy()
  })

  describe('отображает данные уведомления:', () => {
    test('текст.', () => {
      const wrapper = shallowMount(Component, {
        localVue,
        propsData: {
          notification,
        },
        stubs,
      })
      expect(wrapper.find('[data-testid="text"]').text())
        .toContain(notification.text)
    })

    test('заголовок.', () => {
      const wrapper = shallowMount(Component, {
        localVue,
        propsData: {
          notification,
        },
        stubs,
      })
      expect(wrapper.find('[data-testid="title"]').text())
        .toContain(notification.title)
    })

    test('дату.', () => {
      const wrapper = shallowMount(Component, {
        localVue,
        propsData: {
          notification,
        },
        stubs,
      })
      expect(wrapper.find('[data-testid="date"]').text())
        .toContain('10.10.2000')
    })

    test('изображение.', () => {
      const wrapper = shallowMount(Component, {
        localVue,
        propsData: {
          notification,
        },
        stubs,
      })
      expect(wrapper.find('[data-testid="image"]').text())
        .toEqual(notification.image)
    })
  })

  test('корректно отображается.', () => {
    const wrapper = shallowMount(Component, {
      localVue,
      propsData: {
        notification,
      },
      stubs,
    })
    expect(wrapper.element).toMatchSnapshot()
  })

  test('имеет прозрачный фон, если сообщение прочитано.', () => {
    const wrapper = shallowMount(Component, {
      localVue,
      propsData: {
        notification,
      },
      stubs,
    })
    expect(wrapper.attributes().style).toContain('background-color: transparent;')
  })

  test('имеет цветной фон, если сообщение непрочитано.', () => {
    const wrapper = shallowMount(Component, {
      localVue,
      propsData: {
        notification: unseenNotification,
      },
      stubs,
    })
    expect(wrapper.attributes().style).toContain(
      `background-color: ${notification.color.rgba(0.2)};`,
    )
  })
})
