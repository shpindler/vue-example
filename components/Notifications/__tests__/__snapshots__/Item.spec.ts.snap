// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Notification корректно отображается. 1`] = `
<vue-row-stub
  class="notification"
  style="background-color: transparent;"
  tag="article"
>
  <vue-col-stub
    align-self="center"
    class="notification__section"
    cols="auto"
    tag="section"
  >
    <vue-icon-stub
      data-testid="image"
      size="30"
      style="color: rgb(255, 192, 33);"
    >
      
      error
    
    </vue-icon-stub>
  </vue-col-stub>
   
  <vue-col-stub
    class="notification__section"
    tag="section"
  >
    <h2
      class="notification__title"
      data-testid="title"
    >
      
      22
      
      <time
        class="notification__date"
        data-testid="date"
      >
        
        10.10.2000
      
      </time>
    </h2>
     
    <div
      class="notification__text"
      data-testid="text"
      lang="ru"
    >
      Тестовое уведомление.
    </div>
  </vue-col-stub>
</vue-row-stub>
`;
